File: apps/web/lib/badge_catalog.ts
import catalog from '../../../shared/badges.json';
import type { BadgeCategory } from './badges';

export type BadgeRule =
  | { type: 'ctx_gte'; field: string; value: number }
  | { type: 'stats_gte'; field: string; value: number }
  | { type: 'stats_array_len_gte'; field: string; value: number }
  | { type: 'stats_includes'; field: string; value: string }
  | { type: 'stats_bool'; field: string }
  | { type: 'custom'; name: string; args?: Record<string, unknown> };

export type BadgeProgress =
  | { type: 'ratio'; field: string; max: number }
  | { type: 'bool' }
  | { type: 'custom'; name: string; args?: Record<string, unknown> };

export type BadgeSpec = {
  id: string;
  label: string;
  desc: string;
  hint: string;
  icon: string;
  category: BadgeCategory;
  interesting: boolean;
  rule: BadgeRule;
  progress: BadgeProgress;
};

export const BADGE_CATALOG_VERSION = catalog.version as number;
export const BADGE_CATEGORY_LABELS = catalog.categories as Record<BadgeCategory, string>;
export const BADGE_CATEGORY_ORDER = catalog.categoryOrder as BadgeCategory[];
export const BADGE_SPECS = catalog.badges as BadgeSpec[];

/** 旧版成就 ID → 统一契约 ID */
export const LEGACY_BADGE_ID_MAP: Record<string, string> = {
  streak7: 'streak_7',
  streak30: 'streak_30',
  books10: 'books_5',
  nt: 'books_27',
  notes10: 'note_first',
  quiz50: 'memory_review',
  ai5: 'xiaoai_first',
};

export function normalizeBadgeId(id: string): string {
  return LEGACY_BADGE_ID_MAP[id] ?? id;
}

export function normalizeUnlockedAtMap(raw: Record<string, number>): Record<string, number> {
  const out: Record<string, number> = {};
  for (const [k, v] of Object.entries(raw)) {
    const id = normalizeBadgeId(k);
    out[id] = out[id] ? Math.min(out[id], v) : v;
  }
  return out;
}

File: apps/mobile/lib/core/badge_catalog.dart
import 'dart:convert';

import 'package:flutter/services.dart';

class BadgeSpec {
  BadgeSpec({
    required this.id,
    required this.label,
    required this.desc,
    required this.hint,
    required this.icon,
    required this.category,
    required this.interesting,
    required this.rule,
    required this.progress,
  });

  final String id;
  final String label;
  final String desc;
  final String hint;
  final String icon;
  final String category;
  final bool interesting;
  final Map<String, dynamic> rule;
  final Map<String, dynamic> progress;

  factory BadgeSpec.fromJson(Map<String, dynamic> json) => BadgeSpec(
        id: json['id'] as String,
        label: json['label'] as String,
        desc: json['desc'] as String,
        hint: json['hint'] as String,
        icon: json['icon'] as String,
        category: json['category'] as String,
        interesting: json['interesting'] as bool? ?? true,
        rule: Map<String, dynamic>.from(json['rule'] as Map),
        progress: Map<String, dynamic>.from(json['progress'] as Map),
      );
}

class BadgeCatalog {
  BadgeCatalog._(this.version, this.categoryLabels, this.categoryOrder, this.specs);

  final int version;
  final Map<String, String> categoryLabels;
  final List<String> categoryOrder;
  final List<BadgeSpec> specs;

  static BadgeCatalog? _cache;

  static Future<BadgeCatalog> load() async {
    if (_cache != null) return _cache!;
    final raw = await rootBundle.loadString('assets/badges.json');
    final json = jsonDecode(raw) as Map<String, dynamic>;
    final cats = (json['categories'] as Map).map(
      (k, v) => MapEntry('$k', '$v'),
    );
    final order = (json['categoryOrder'] as List).cast<String>();
    final specs = (json['badges'] as List)
        .map((e) => BadgeSpec.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
    _cache = BadgeCatalog._(json['version'] as int? ?? 1, cats, order, specs);
    return _cache!;
  }
}

/// 旧版 Mobile 成就 ID → 统一契约 ID
const legacyBadgeIdMap = <String, String>{
  'streak7': 'streak_7',
  'streak30': 'streak_30',
  'books10': 'books_5',
  'nt': 'books_27',
  'notes10': 'note_first',
};

String normalizeBadgeId(String id) => legacyBadgeIdMap[id] ?? id;
